source/Fleet.cpp

```diff
@@ -515,18 +517,24 @@ void Fleet::Place(const System &system, list<shared_ptr<Ship>> &ships, bool carr
 // Do the randomization to make a ship enter or be in the given system.
 const System *Fleet::Enter(const System &system, Ship &ship, const System *source)
 {
-	if(system.Links().empty() || (source && !system.Links().count(source)))
+	bool canEnter = false;
+	for(const System *link : system.Links())
+		if(!ship.GetGovernment()->Restricted(link))
+			canEnter = true;
+			
+	if(!canEnter || system.Links().empty() || (source && !system.Links().count(source)))
 	{
 		Place(system, ship);
 		return &system;
 	}
 
 	// Choose which system this ship is coming from.
-	if(!source)
+	while(!source)
 	{
 		auto it = system.Links().cbegin();
 		advance(it, Random::Int(system.Links().size()));
-		source = *it;
+		if(!ship.GetGovernment()->Restricted(source))
+			source = *it;
```

Comment:
I'd rather only fetch one random number rather than possibly infinite random numbers. Compute a weighted sum (where restricted links have no weight) and then generate a number up to that sum. Select the appropriate link based on the picked number.

(This approach is used by ai to select a planet vs destination system if you want a reference implementation.)